File: py/idkollen_client/models/vipps.py
from __future__ import annotations

from typing import Annotated, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from idkollen_client.models.common import ApiErrorCode


class _Base(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class VippsAuthRequest(_Base):
    redirect_url: Optional[str] = None
    request_ssn: Optional[bool] = None
    request_phone: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    ref_id: Optional[str] = None
    app_callback_uri: Optional[str] = None


class VippsBackchannelAuthRequest(_Base):
    phone: str
    request_ssn: Optional[bool] = None
    request_email: Optional[bool] = None
    request_address: Optional[bool] = None
    callback_url: Optional[str] = None
    ref_id: Optional[str] = None


class VippsPending(_Base):
    status: Literal["PENDING"]
    id: str
    ref_id: Optional[str] = None
    url: Optional[str] = None


class VippsCompleted(_Base):
    status: Literal["COMPLETED"]
    id: str
    ref_id: Optional[str] = None
    ssn: str
    name: str
    given_name: str
    surname: str
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    birth_date: Optional[str] = None
    pid: Optional[str] = None
    bank_id: Optional[str] = None


class VippsFailed(_Base):
    status: Literal["FAILED"]
    id: str
    ref_id: Optional[str] = None
    error: ApiErrorCode


VippsStatus = Annotated[
    VippsPending | VippsCompleted | VippsFailed,
    Field(discriminator="status"),
]

File: csharp/src/Idkollen.Client/Models/Vipps.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Idkollen.Client.Models;

public sealed record VippsAuthRequest
{
    public string? RedirectUrl { get; init; }
    public bool? RequestSsn { get; init; }
    public bool? RequestPhone { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? RefId { get; init; }
    public string? AppCallbackUri { get; init; }
}

public sealed record VippsBackchannelAuthRequest(string Phone)
{
    public bool? RequestSsn { get; init; }
    public bool? RequestEmail { get; init; }
    public bool? RequestAddress { get; init; }
    public string? CallbackUrl { get; init; }
    public string? RefId { get; init; }
}

public interface IVippsStatus { }

public sealed record VippsPending(string Id, string? RefId, string? Url) : IVippsStatus;

public sealed record VippsCompleted(
    string Id,
    string? RefId,
    string Ssn,
    string Name,
    string GivenName,
    string Surname,
    string? Phone,
    string? Email,
    string? Address,
    string? BirthDate,
    string? Pid,
    string? BankId
) : IVippsStatus;

public sealed record VippsFailed(string Id, string? RefId, string Error) : IVippsStatus;

internal sealed class VippsStatusConverter : JsonConverter<IVippsStatus>
{
    public override IVippsStatus Read(ref Utf8JsonReader reader, Type _, JsonSerializerOptions options)
    {
        using var doc = JsonDocument.ParseValue(ref reader);
        var status = doc.RootElement.GetProperty("status").GetString();
        var raw = doc.RootElement.GetRawText();
        return status switch
        {
            "PENDING" => JsonSerializer.Deserialize<VippsPending>(raw, options)!,
            "COMPLETED" => JsonSerializer.Deserialize<VippsCompleted>(raw, options)!,
            "FAILED" => JsonSerializer.Deserialize<VippsFailed>(raw, options)!,
            _ => throw new JsonException($"Unknown vipps status: {status}"),
        };
    }

    public override void Write(Utf8JsonWriter w, IVippsStatus v, JsonSerializerOptions o)
        => throw new NotSupportedException();
}
